flakeseries_table.go

```diff
@@ -0,0 +1,130 @@
+package gocassa
+
+import (
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/mattheath/base62"
+	"github.com/mattheath/kala/bigflake"
+	"github.com/mattheath/kala/util"
+)
+
+// we have to put the timestamp from the flake ID into a field so that we can
+// execute queries with it
+const flakeTimestampFieldName = "flake_created"
+
+type flakeSeriesT struct {
+	Table
+	bucketSize time.Duration
+}
+
+func (o *flakeSeriesT) Set(v interface{}) Op {
+	m, ok := toMap(v)
+	if !ok {
+		panic("Can't set: not able to convert")
+	}
+	id, ok := m["Id"].(string)
+	if !ok {
+		panic("Id field is not present or is not a string")
+	}
+
+	timestamp, err := flakeToTime(id)
+	if err != nil {
+		return errOp{err: err}
+	}
+
+	m[flakeTimestampFieldName] = timestamp
+	m[bucketFieldName] = o.bucket(timestamp.Unix())
+
+	return o.Table.Set(m)
+}
+
+func (o *flakeSeriesT) Update(id string, m map[string]interface{}) Op {
+	timestamp, err := flakeToTime(id)
+	if err != nil {
+		return errOp{err: err}
+	}
+	bucket := o.bucket(timestamp.Unix())
+
+	return o.Where(Eq(bucketFieldName, bucket), Eq(flakeTimestampFieldName, timestamp), Eq("Id", id)).Update(m)
+}
+
+func (o *flakeSeriesT) Delete(id string) Op {
+	timestamp, err := flakeToTime(id)
+	if err != nil {
+		return errOp{err: err}
+	}
+	bucket := o.bucket(timestamp.Unix())
+
+	return o.Where(Eq(bucketFieldName, bucket), Eq(flakeTimestampFieldName, timestamp), Eq("Id", id)).Delete()
+}
+
+func (o *flakeSeriesT) Read(id string, pointer interface{}) Op {
+	timestamp, err := flakeToTime(id)
+	if err != nil {
+		return errOp{err: err}
+	}
+	bucket := o.bucket(timestamp.Unix())
+	return o.Where(Eq(bucketFieldName, bucket), Eq(flakeTimestampFieldName, timestamp), Eq("Id", id)).ReadOne(pointer)
+}
+
+func (o *flakeSeriesT) List(startTime, endTime time.Time, pointerToASlice interface{}) Op {
+	buckets := o.buckets(startTime, endTime)
+	return o.Where(In(bucketFieldName, buckets...), GTE(flakeTimestampFieldName, startTime), LT(flakeTimestampFieldName, endTime)).Read(pointerToASlice)
+}
+
+// ListSince queries the flakeSeries for the items after the specified ID but within the time window,
+// if the time window is zero then it lists up until 5 minutes in the future
+func (o *flakeSeriesT) ListSince(id string, window time.Duration, pointerToASlice interface{}) Op {
+	startTime, err := flakeToTime(id)
+	if err != nil {
+		return errOp{err: err}
+	}
+
+	var endTime time.Time
+
+	if window == 0 {
+		// no window set - so go up until 5 mins in the future
+		endTime = time.Now().Add(5 * time.Minute)
```

Comment:
All timestamps extracted from Flake IDs, and all those used in queries (like here) should probably be converted to UTC.